issue9-federated-sparql-queries

RhythmSyed opened the pull request:
Added remote querying capability using federated queries and SERVICE
Added a notebook example that includes variations of remote service queries and an example widget

RhythmSyed commented:
Issue: #9

zwelz3 commented:
I don't think we want `.DS_Store` in source control. Let's add to the gitignore

zwelz3 commented:
There is a typo in `examples/RemoteQuery_Example.ipynb`. The triple pattern should be `?s ?p ?o`, but in several locations the typo is `?s ?o ?o`, which causes the results to be wrong.

zwelz3 commented:
For the federated query example can we show an example of query data across the local (in memory ) graph and a remote graph? 

Something like below (even though it doesn't actually work). Trying to illustrate the concept. 
```python
from rdflib import namespace, Graph, Literal, URIRef

WD = namespace.Namespace("https://www.wikidata.org/wiki/")

graph = Graph()
graph.add((WD.Q28792126, WD.example, Literal("Example")))
query_str = """
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX wikibase: <http://wikiba.se/ontology#>
PREFIX p: <http://www.wikidata.org/prop/>
PREFIX ps: <http://www.wikidata.org/prop/statement/>
PREFIX pq: <http://www.wikidata.org/prop/qualifier/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX bd: <http://www.bigdata.com/rdf#>

SELECT ?p ?o
WHERE {
    wd:Q28792126 ?p ?o .

    service <https://query.wikidata.org/> 
    {
        #Cats
        SELECT ?p ?o 
        WHERE 
        {
          BIND(wikibase:label AS ?p)
          wd:Q28792126 wdt:P31 wd:Q146.
          service wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE],en". }
        } 
    }
}
"""
res = graph.query(query_str)
list(res)
```

zwelz3 commented:
The example for multiple remotes uses some weird limits. Lets make the inner selects limit=4 and the outer select limit=4.

zwelz3 commented:
Let's remove the remote query widget completely. For now we will just require the users of the regular query widget to use nested select with service calls. In the future, I think we could add a remote query widget that wraps all queries, but we don't want to overcomplicate at the moment.

zwelz3 commented:
Can we add a little more narrative around the notebook code? It would be valuable to explain why these queries are important (e.g. for data integration and distributed knowledge).

zwelz3 commented:
@nrbgt are the changes to `examples/Tab_App_Example.ipynb` in the diff a result of the line endings? Is this issue on the windows developers' side or on Rhythm's?

zwelz3 commented:
Let's break up your `examples/RemoteQuery_Example.ipynb` notebook into two notebooks: `RemoteQuery_Example` and `FederatedQuery_Example` and add both to the index under `Query Examples` as `Remote query example` and `Federated query example`.

zwelz3 commented:
We will need some basic tests to ensure that the code you added is working as expected. I'm thinking at least:
1. test for the service wrapper
2. test for no service wrapper (expect fail)
3. test for remote query (check results)
4. test for federated query (check results)

These can be pretty simple for now.

zwelz3 commented:
Good work so far. Once you get done with these changes, please update your description with a more comprehensive discussion of the work performed and how to validate the PR (e.g. `doit all` and run xyz notebook).

RhythmSyed commented:
Binder: https://mybinder.org/v2/gh/rhythmsyed/ipyradiant/remotequery?urlpath=lab/tree/index.ipynb

RhythmSyed commented:
To run:
- doit lab
- run RemoteQuery_Example.ipynb
- run FederatedQuery_Example.ipynb

To run unit tests:
- pytest examples/tests

Changes:
- Added logging for RDFLib service patch, can be turned off from user input ('CRITICAL')
- Added example of combining local graph + remote graph via federated service query
- Added a broken example of nested service calls that is not supported in RDFLib
- Added more narrative to RemoteQuery and FederatedQuery notebooks
- Added pytests for service_patch and query examples
- Added utils_helper in /query which fixed circular import issues. 
- service_patch can be called by ipyradiant.service_patch_rdflib, passing in a query string
- logger level can be called by ipyradiant.set_logger_level, passing in a logger level such as 'DEBUG' or 'CRITICAL'
- Tested ipyradiant widgets with service calls which successfully get routed to patch then return the fixed query string

sanbales commented:
The `FederatedQuery_Example.ipynb` is crashing the kernel on the 12th cell:

```python
res = graph.query(query_str)
list(res)
```

RhythmSyed commented:
Updated November 20th:

To run:
- doit lab / jupyter lab
- run RemoteQuery_Example.ipynb
- run FederatedQuery_Example.ipynb

To run unit tests:
- pytest examples/tests

Changes:
- Removed remote_query module
- Added persistent tests for external APIs
- Reformatted unit tests to use pytest
- Added testing as a step in git CI
- Reformatted utils_helper to utils
- Removed set_logger_level function
- In the remotequery notebook, added an example to show the query widget using service patch
- Restructured notebooks to use working endpoints for sparql queries and added more narrative
- In the remotequery notebook, removed redundant examples
- In the remotequery notebook, fixed nested query example and warning from other examples
- In the federated query notebook, fixed the in-memory and remote graph example
- Added another federated example that shows interesting results
- Added two remote endpoint querying example

nrbgt commented on .github/workflows/ciV0.yml:
```diff
@@ -76,6 +76,10 @@ jobs:
           name: ipyradiant dist ${{ github.run_number }}
           path: ./dist
 
+      -name: run pytests
```
add space, e.g. `- name: run pytest`? or just omit `name` altogether (it's short and self-explanatory)

zwelz3 commented:
Last couple of comments:
* examples/RemoteQuery_Example.ipynb
  * [x] Sort the package imports
  * [x] `This patch is automatically removed for release>5.0.0` its not really removed. 
  * [x] Also, logger is not printing the info for me in the notebook?
  * [x] `Once rdflib is updated, the INFO message is removed by setting the logger_level to WARNING or higher.` Lets remove this cell. I don't think this is the "right" way to correct the warning.
  * [x]  Your `Widget Example` should probably prepopulate a service statement

* ipyradiant/query/utils.py
  * [x] There is probably a better way to deal with deprecating the service patch with rdflib version, but for now, maybe we can be a bit more explicit.
```python
def service_patch_rdflib(query_str):
    # check for rdflib version, if <=5.0.0 throw warning
    version = rdflib.__version__
    major_version_num = int(version.split(".")[0])

    # if version > 5, warn users ipyradiant needs to be updated
    if major_version_num <= 5 and "SERVICE" in query_str:
        query_str = query_str.replace("SERVICE", "service")
        logger.info(
            "SERVICE found in query. RDFlib currently only supports `service`. "
            "To be fixed in the next release>5.0.0"
        )
    elif major_version_num > 5:
        logger.info("Service patch for ipyradiant should be removed.")

    return query_str
```


* examples/FederatedQuery_Example.ipynb
  * [x] sort imports
  * [x] `Federated queries allows the capability to gather knowledge from distributed sources into one aggregated knowledge graph.` One small correction, they are not (necessarily) aggregated into a knowledge graph, but they are aggregated into a single query result. 
  * [x] `Here we create a small triple as an in-memory graph` -> Here we create a _single_ triple _that will be used_ as...
  * [x] `In this example, we are supplementing our local graph with data` -> In this _next_ example...
  * [x] I think your `Query Two Remote Endpoints` should be `Broken Example: Query Two....` to drive home the point that it doesn't work. Also I think the query should be:
```python
query_str = """
    SELECT ?s
    WHERE {
        {
            service <https://query.wikidata.org/sparql> 
            {
                SELECT DISTINCT ?s
                WHERE {?s ?p ?o}
                LIMIT 4                
            }
       }
          UNION
       {
             service <https://query.wikidata.org/sparql> 
            {
                SELECT DISTINCT ?s
                WHERE {?s ?p ?o}
                LIMIT 4   
                OFFSET 4
            }
       }
    }
"""
```
